qiita_db/metadata_template.py

```diff
@@ -1590,6 +1590,25 @@ def create(cls, md_template, raw_data, study, data_type,
             conn_handler.execute(
                 "DELETE FROM qiita.prep_template where "
                 "{0} = %s".format(cls._id_column), (prep_id,))
+
+            # Check if sample IDs present here but not in sample template
+            sql = ("SELECT sample_id from qiita.required_sample_info WHERE "
+                   "study_id = %s")
+            # Get list of study sample IDs, prep template study IDs,
+            # and their intersection
+            study_samples = set(s[0] for s in conn_handler.execute_fetchall(
+                sql, [study.id]))
+            prep_samples = set(md_template.index.values)
+            both_samples = study_samples.intersection(prep_samples)
+            # filter prep template to just samples available in the study
+            md_template = md_template.loc[both_samples]
+            if len(both_samples) != len(prep_samples):
+                prep_samples.difference_update(both_samples)
+                raise QiitaDBExecutionError(
+                    'Samples found in prep template but not sample template: '
+                    '%s' % ', '.join(prep_samples))
```

Comment:
The code confuses me slightly, and I don't think it can be simplified:

```python
unknown_sample_ids = prep_samples - study_samples
if unknown_sample_ids:
    # Raise the QiitaDBExecutionError as before
```